File: python/test/testsupport/stub_metadata_plugin.py
from dataclasses import dataclass
from typing import Dict, Union

from models.document import Document
from plugins.md2html_plugin import Md2HtmlPlugin, MetadataProcessingResult


@dataclass(frozen=True)
class MarkerSpec:
    marker: str
    only_at_page_start: bool = False

    @staticmethod
    def anywhere(marker: str) -> 'MarkerSpec':
        return MarkerSpec(marker, False)

    @staticmethod
    def at_page_start(marker: str) -> 'MarkerSpec':
        return MarkerSpec(marker, True)


class StubMetadataPlugin(Md2HtmlPlugin):

    def __init__(self, *marker_specs: MarkerSpec):
        super().__init__()
        self._handler_infos = [
            (self, spec.marker, spec.only_at_page_start) for spec in marker_specs
        ]

    def accept_data(self, data):
        raise NotImplementedError("Not used in stub plugin tests")

    def is_blank(self) -> bool:
        return not self._handler_infos

    def page_metadata_handlers(self):
        return self._handler_infos

    def accept_page_metadata(self, doc: Document, marker: str, metadata: str,
                             metadata_section: str,
                             visited_markers: Union[Dict[str, None], None] = None,
                             phase: int = 1, data_from_prev_phase=None
                             ) -> MetadataProcessingResult:
        return MetadataProcessingResult(metadata)

File: java/src/test/java/world/md2html/plugins/testsupport/StubMetadataPlugin.java
package world.md2html.plugins.testsupport;

import com.fasterxml.jackson.databind.JsonNode;
import world.md2html.options.argfile.ArgFileParseException;
import world.md2html.options.model.Document;
import world.md2html.pagemetadata.MetadataProcessingResult;
import world.md2html.plugins.Md2HtmlPlugin;
import world.md2html.plugins.PageMetadataHandler;
import world.md2html.plugins.PageMetadataHandlerInfo;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

/**
 * Test-only plugin that registers configurable metadata markers without production plugin config.
 */
public class StubMetadataPlugin implements Md2HtmlPlugin, PageMetadataHandler {

    public static final class MarkerSpec {
        private final String marker;
        private final boolean onlyAtPageStart;

        public MarkerSpec(String marker, boolean onlyAtPageStart) {
            this.marker = marker;
            this.onlyAtPageStart = onlyAtPageStart;
        }

        public static MarkerSpec anywhere(String marker) {
            return new MarkerSpec(marker, false);
        }

        public static MarkerSpec atPageStart(String marker) {
            return new MarkerSpec(marker, true);
        }
    }

    private final List<PageMetadataHandlerInfo> handlerInfos;

    public StubMetadataPlugin(MarkerSpec... markerSpecs) {
        List<PageMetadataHandlerInfo> infos = new ArrayList<>();
        for (MarkerSpec spec : markerSpecs) {
            infos.add(new PageMetadataHandlerInfo(this, spec.marker, spec.onlyAtPageStart));
        }
        this.handlerInfos = Collections.unmodifiableList(infos);
    }

    @Override
    public void acceptData(JsonNode data) throws ArgFileParseException {
        throw new UnsupportedOperationException("Not used in stub plugin tests");
    }

    @Override
    public boolean isBlank() {
        return handlerInfos.isEmpty();
    }

    @Override
    public List<PageMetadataHandlerInfo> pageMetadataHandlers() {
        return handlerInfos;
    }

    @Override
    public MetadataProcessingResult acceptPageMetadata(Document document, String marker,
            String metadata, String metadataSection, Set<String> visitedMarkers
    ) throws PageMetadataException {
        return MetadataProcessingResult.immediate(metadata);
    }
}
